GitHub warning annotations

timkrins opened the pull request:
I don't actually intend for this PR to be merged as-is, but more created it as a point of discussion (and for my own use, as it solves my immediate use-case).

Adds a report generator that prints GitHub style annotations to stdout.

Right now, only `warning` is supported - I couldn't immediately determine how to pass a command line switch to the generator without changing the base class.

Screenshot of `diff-cover coverage.lcov --github-warning-annotations` running in GitHub actions
|--|
|<img width="586" alt="Screenshot 2025-01-12 at 10 36 30" src="https://github.com/user-attachments/assets/332e3bcf-f7f0-4f7b-9b98-4cc94b790d13" />|
|<img width="586" alt="Screenshot 2025-01-12 at 11 15 53" src="https://github.com/user-attachments/assets/5e623e5b-d6f2-4af8-898d-787c0ba18b52" />|

wyardley commented on diff_cover/diff_cover_tool.py:
```diff
@@ -92,6 +94,13 @@ def parse_coverage_args(argv):
     )
 
+    parser.add_argument(
+        "--github-warning-annotations",
```
given the idea of moving to a plugin type framework, maybe this should be implemented as a top level `--format` flag, similar to some other programs, so that it's possible to do `--format github-native` or `--format xyz`?

There would have to be discussion then about whether multiple formats are allowed (and if so, how)... for example, would it allow `--format json,markdown,github-native` or multiple instances of the same flag, something like `--format github-native --format markdown:foo.md` (the latter case allowing specifying the format and filename together)? If it's too invasive to take over the markdown / json / etc. reports, could at least have a single flag to control all stdout / stderr reporting.

Bachmann1234 replied:
Yep, I like this

Bachmann1234 replied:
I will say if we move to this style then we should have it support the existing formats as well as the new one. 

We kinda have to keep the existing flags to avoid breaking compatibility, but if we are going to be adding more then moving to something like this makes sense

timkrins replied:
I actually started down the path of attempting to pass something like `--github-annotations=warning` or `--github-annotations=error`.
But I saw that customizing the context that is passed to jinja template for specific template generators wasn't so easy to adapt.

Bachmann1234 replied:
@timkrins if you wanted to have that context could that be achieved by adding some kind of `context` parameter to the __init__ of BaseReportGenerator?

timkrins replied:
Maybe - I'll try have a go later in the week

kingbuzzman replied:
This has changed significantly, and adding `--format github` is now possible.

Bachmann1234 commented:
So at a high level this seems like a useful feature. I have a hard time thinking about new diff-cover/diff-quality features since I dont really *use* these tools anymore. I just maintain the project.

But it looks like Github annotations introduce some level of customizability (whats a warning, whats an error, whats a message). My gut is the approach taken here "Everything is a warning" seems reasonable. However, if others start using it I suspect the desire for customization will come up shortly after.  

I think im happy to merge in a "everything is a warning" pr. Do yall think there will be much demand for further customization? Do you think we could get away with "look, the flag will do the basics but if you have more complex demands that config needs to be defined in some config file"?

wyardley commented:
For the use case like sqlfluff, both errors and warnings would be ideal.

Being able to add code suggestions is also useful, though I suspect most people will use a tool like [reviewdog](https://github.com/reviewdog/reviewdog) instead for that.

kingbuzzman commented:
Guess who had this same idea today.... [independently] ME! I'll help out, I would like to see this feature soon.

timkrins commented:
@kingbuzzman awesome regarding the format arguments. I'm a bit slammed atm but I will try and extract some time to rebase!

kingbuzzman commented:
No rebase needed, just in github click on the "resolve conflicts" button 😉

timkrins commented:
I have rebased on main, and added the annotation format - although right now it is a TODO as I am still a bit unclear on an elegant way to pass the value down to the formatter and then template context.

kingbuzzman commented:
@timkrins hey, thanks for getting to this, can you sync with `main`?

kingbuzzman commented:
Thank you!

timkrins commented on diff_cover/diff_cover_tool.py:
```diff
@@ -266,9 +268,18 @@ def generate_coverage_report(
         markdown_report = report_formats["markdown"] or MARKDOWN_REPORT_DEFAULT_PATH
         reporter = MarkdownReportGenerator(coverage, diff)
-        with open(markdown_report, "wb") as output_file:
+        with open_file(markdown_report, "wb") as output_file:
```
Removed.

kingbuzzman replied:
Fixed here: https://github.com/Bachmann1234/diff_cover/pull/516

timkrins commented on .github/workflows/verify.yaml:
```diff
@@ -97,7 +97,7 @@ jobs:
       - name: Complete coverage
         run: |
-          poetry run diff-cover coverage.xml --include-untracked
+          poetry run diff-cover coverage.xml --include-untracked --format github-annotations:error
```
I will change it to warning.

kingbuzzman replied:
There is no "default" it will crash. This is a bigger bug that has little to do with your PR

Bachmann1234 commented:
Alrighty! Lets go ahead and merge/release this. Thank you for the PR @timkrins  and thanks for the review support @kingbuzzman

Bachmann1234 commented:
Released! Check out https://pypi.org/project/diff-cover/9.6.0/
